README.md
# SBT Encrypt

[![style: very good analysis][very_good_analysis_badge]][very_good_analysis_link]
[![License: MIT][license_badge]][license_link]

Encrypt with argon2 + chacha20.
## Usage
```dart
final encprypted = await encrypt('test', 'password')
final decrypted = await decrypt(encrypted, 'password')
```

[dart_install_link]: https://dart.dev/get-dart
[github_actions_link]: https://docs.github.com/en/actions/learn-github-actions
[license_badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license_link]: https://opensource.org/licenses/MIT
[logo_black]: https://raw.githubusercontent.com/VGVentures/very_good_brand/main/styles/README/vgv_logo_black.png#gh-light-mode-only
[logo_white]: https://raw.githubusercontent.com/VGVentures/very_good_brand/main/styles/README/vgv_logo_white.png#gh-dark-mode-only
[mason_link]: https://github.com/felangel/mason
[very_good_analysis_badge]: https://img.shields.io/badge/style-very_good_analysis-B22C89.svg
[very_good_analysis_link]: https://pub.dev/packages/very_good_analysis
[very_good_coverage_link]: https://github.com/marketplace/actions/very-good-coverage
[very_good_ventures_link]: https://verygood.ventures
[very_good_ventures_link_light]: https://verygood.ventures#gh-light-mode-only
[very_good_ventures_link_dark]: https://verygood.ventures#gh-dark-mode-only
[very_good_workflows_link]: https://github.com/VeryGoodOpenSource/very_good_workflows


CHANGELOG.md
# 0.1.4

-chore: update cryptography

# 0.1.3

-fix: bugs fixed.
# 0.1.2

- feat: change to use chacha20.
# 0.1.0+1

- feat: initial commit 🎉


test/src/aes_dart_test.dart
// Copyright (c) 2022, Very Good Ventures
// https://verygood.ventures
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:aes_dart/aes_dart.dart';
// ignore_for_file: prefer_const_constructors
import 'package:test/test.dart';

void main() {
  group('AesDart', () {
    test('can encode', () async {
      final encrypted = await aesEncrypt('test', '121212');
      final decrypted = await aesDecrypt(encrypted, '121212');
      expect('test', decrypted);
    });

    test('decypt data', () async {
      const data =
          '1Gyw5Al31seI7s/V7Ma5gQ3K4/lxmuqpr4OHG0FPAfxPRkraylz8I7HvA1/h5E4g';
      final decrypted = await aesDecrypt(data, '121212');
      expect('test', decrypted);
    });
  });
}


test/src/sbt_encrypt_test.dart
// Copyright (c) 2022, Very Good Ventures
// https://verygood.ventures
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'dart:convert';
import 'dart:typed_data';

import 'package:sbt_encrypt/src/sbt_encrypt.dart';
// ignore_for_file: prefer_const_constructors
import 'package:test/test.dart';

void main() {
  group('sbt encrypt test', () {
    test('argon2 derive', () async {
      const password = '121212';
      const salt = [158, 82, 111, 126, 245, 222, 179, 76, 133, 117, 16, 149];
      final result = await argon(password, Uint8List.fromList(salt));
      expect(
        base64.encode(result),
        'F30DpIwR02Iz7/A+61quxaPTQnqNvcNbFlSs83wZfT4=',
      );
    });
    test('can encode', () async {
      final encrypted = await encrypt('test', '121212');
      final decrypted = await decrypt(encrypted, '121212');
      expect('test', decrypted);
    });

    test('decypt data', () async {
      const data = 'z7E6AV2FtDqvHB5Ral9OIQ==';
      final decrypted = await decrypt(data, '121212');
      expect('test', decrypted);
    });
  });
}


lib/src/aes_dart.dart
// Copyright (c) 2022, Very Good Ventures
// https://verygood.ventures
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.
import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';
import 'package:encrypt/encrypt.dart';

/// encrypt data
Future<String> aesEncrypt(String plainText, String password) async {
  final iv = IV.fromSecureRandom(16);
  final salt = IV.fromSecureRandom(16).bytes;
  final pbkdf2 = Pbkdf2(
    macAlgorithm: Hmac.sha256(),
    iterations: 1,
    bits: 512,
  );
  final newSecretKey = await pbkdf2.deriveKey(
    secretKey: SecretKey(ascii.encode(password)),
    nonce: salt,
  );
  final key = await newSecretKey.extractBytes();
  final encrypter = Encrypter(
    AES(
      Key(Uint8List.fromList(key.sublist(0, 32))),
      mode: AESMode.cbc,
    ),
  );
  final encrypted = encrypter.encrypt(plainText, iv: iv);
  final packaged = iv.bytes + salt + encrypted.bytes;
  return base64.encode(packaged);
}

/// decrypted data
Future<String> aesDecrypt(String input, String password) async {
  final unpackaged = base64.decode(input);
  final iv = IV(unpackaged.sublist(0, 16));
  final salt = unpackaged.sublist(16, 32);
  final data = unpackaged.sublist(32);
  final pbkdf2 = Pbkdf2(
    macAlgorithm: Hmac.sha256(),
    iterations: 1,
    bits: 512,
  );
  final newSecretKey = await pbkdf2.deriveKey(
    secretKey: SecretKey(ascii.encode(password)),
    nonce: salt,
  );
  final key = await newSecretKey.extractBytes();
  final encrypter = Encrypter(
    AES(
      Key(Uint8List.fromList(key.sublist(0, 32))),
      mode: AESMode.cbc,
    ),
  );
  final encrypted = encrypter.decrypt(Encrypted(data), iv: iv);
  return encrypted;
}


lib/src/sbt_encrypt.dart
// Copyright (c) 2022, Very Good Ventures
// https://verygood.ventures
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.
import 'dart:convert';
import 'dart:typed_data';

import 'package:argon2/argon2.dart';
import 'package:cryptography/cryptography.dart';
import 'package:encrypt/encrypt.dart';
import 'package:sbt_encrypt/src/chacha20.dart';

/// Password derive
Future<List<int>> argon(String password, Uint8List salt) async {
  final parameters = Argon2Parameters(
    Argon2Parameters.ARGON2_d,
    salt,
    iterations: 2,
    memoryPowerOf2: 5,
  );
  final argon2 = Argon2BytesGenerator()..init(parameters);
  final passwordBytes = parameters.converter.convert(password);
  final result = Uint8List(32);
  argon2.generateBytes(passwordBytes, result, 0, result.length);
  return result;
}

/// encrypt data
Future<String> encrypt(String plainText, String password) async {
  // final iv = IV.fromSecureRandom(16);
  final salt = IV.fromSecureRandom(12).bytes;
  final key = await argon(password, salt);
  final algorithm = SbtChacha20(macAlgorithm: Hmac.sha256());
  final secretBox = await algorithm.encrypt(
    ascii.encode(plainText),
    secretKey: SecretKey(key.sublist(0, 32)),
    nonce: salt,
  );
  return base64.encode(secretBox.nonce + secretBox.cipherText);
}

/// decrypted data
Future<String> decrypt(String input, String password) async {
  final unpackaged = base64.decode(input);
  final salt = unpackaged.sublist(0, 12);
  final data = unpackaged.sublist(12);
  final secretBox = SecretBox(data, nonce: salt, mac: Mac([]));
  final key = await argon(password, salt);
  final algorithm = SbtChacha20(macAlgorithm: Hmac.sha256());
  final encrypted = await algorithm.decrypt(
    secretBox,
    secretKey: SecretKey(key.sublist(0, 32)),
  );
  return ascii.decode(encrypted);
}


lib/src/chacha20.dart
// Copyright 2019-2020 Gohilla Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';
import 'package:cryptography/dart.dart';
import 'package:cryptography/src/utils.dart';

/// [Chacha20] implemented in pure Dart.
///
/// ## Recommendation
/// Usually you should use [Cryptography.chacha20]:
/// ```
/// final cipher = Cryptography.instance.chacha20()
/// ```
class SbtChacha20 extends Chacha20 {
  static const int _rounds = 20;

  @override
  final MacAlgorithm macAlgorithm;

  const SbtChacha20({
    required this.macAlgorithm,
  }) : super.constructor();

  @override
  Future<List<int>> decrypt(SecretBox secretBox,
      {required SecretKey secretKey,
      List<int> aad = const <int>[],
      int keyStreamIndex = 0,
      Uint8List? possibleBuffer}) async {
    if (keyStreamIndex < 0) {
      throw ArgumentError.value(
        keyStreamIndex,
        'keyStreamIndex',
        'Must be non-negative',
      );
    }
    if (secretBox.nonce.length != 12) {
      throw ArgumentError.value(
        secretBox,
        'secretBox',
        'Nonce must have 12 bytes',
      );
    }
    final secretKeyData = await secretKey.extract();
    if (secretKeyData.bytes.length != 32) {
      throw ArgumentError.value(
        secretKey,
        'secretKey',
        'Must have 32 bytes',
      );
    }
    final macAlgorithm = this.macAlgorithm;
    if (macAlgorithm is DartChacha20Poly1305AeadMacAlgorithm) {
      if (keyStreamIndex != 0) {
        throw ArgumentError.value(
          keyStreamIndex,
          'keyStreamIndex',
          'Must be zero when AEAD is used',
        );
      }
      keyStreamIndex = 64;
    }

    // Authenticate
    // await secretBox.checkMac(
    //   macAlgorithm: macAlgorithm,
    //   secretKey: secretKey,
    //   aad: aad,
    // );

    return _xorSync(
      secretBox.cipherText,
      secretKey: secretKeyData,
      nonce: secretBox.nonce,
      aad: aad,
      keyStreamIndex: keyStreamIndex,
    );
  }

  @override
  Future<SecretBox> encrypt(List<int> clearText,
      {required SecretKey secretKey,
      List<int>? nonce,
      List<int> aad = const <int>[],
      int keyStreamIndex = 0,
      Uint8List? possibleBuffer,}) async {
    if (keyStreamIndex < 0) {
      throw ArgumentError.value(
        keyStreamIndex,
        'keyStreamIndex',
        'Must be non-negative',
      );
    }
    nonce ??= newNonce();
    if (nonce.length != 12) {
      throw ArgumentError.value(
        nonce,
        'nonce',
        'Nonce must have 12 bytes',
      );
    }
    final macAlgorithm = this.macAlgorithm;
    if (macAlgorithm is DartChacha20Poly1305AeadMacAlgorithm) {
      if (keyStreamIndex != 0) {
        throw ArgumentError.value(
          keyStreamIndex,
          'keyStreamIndex',
          'Must be zero',
        );
      }
      keyStreamIndex = 64;
    }

    final secretKeyData = await secretKey.extract();
    if (secretKeyData.bytes.length != 32) {
      throw ArgumentError.value(
        secretKey,
        'secretKey',
        'length must be 32 bytes',
      );
    }
    final cipherText = _xorSync(
      clearText,
      secretKey: secretKeyData,
      nonce: nonce,
      aad: aad,
      keyStreamIndex: keyStreamIndex,
    );
    final mac = await macAlgorithm.calculateMac(
      cipherText,
      secretKey: secretKey,
      nonce: nonce,
      aad: aad,
    );
    return SecretBox(
      cipherText,
      nonce: nonce,
      mac: mac,
    );
  }

  Uint8List _xorSync(
    List<int> clearText, {
    required SecretKeyData secretKey,
    required List<int> nonce,
    List<int> aad = const <int>[],
    int keyStreamIndex = 0,
  }) {
    if (keyStreamIndex < 0) {
      throw ArgumentError.value(keyStreamIndex, 'keyStreamIndex');
    }
    if (secretKey.bytes.length != 32) {
      throw ArgumentError.value(
        secretKey,
        'secretKey',
      );
    }
    if (nonce.length != 12) {
      throw ArgumentError.value(nonce, 'nonce', 'Nonce must have 12 bytes');
    }
    final secretKeyBytes = secretKey.bytes;
    final initialState = Uint32List(16);
    initializeChacha(
      initialState,
      key: secretKeyBytes,
      nonce: nonce,
      keyStreamIndex: keyStreamIndex,
    );
    final skipped = keyStreamIndex % 64;
    final stateAsUint32List = Uint32List(
      (skipped + clearText.length + 63) ~/ 64 * 16,
    );
    final stateAsUint8List = Uint8List.view(stateAsUint32List.buffer);
    stateAsUint8List.setAll(skipped, clearText);
    for (var i = 0; i < stateAsUint32List.length; i += 16) {
      chachaRounds(
        stateAsUint32List,
        i,
        initialState,
        rounds: _rounds,
      );
      initialState[12]++;
    }
    return Uint8List.view(
      stateAsUint32List.buffer,
      skipped,
      clearText.length,
    );
  }

  static void chachaRounds(
    Uint32List state,
    int si,
    Uint32List initialState, {
    required int rounds,
    bool addAndXor = true,
  }) {
    // -------------------------------------------------------------------------
    // Step 1: Initialize
    // -------------------------------------------------------------------------
    var v0 = initialState[0],
        v1 = initialState[1],
        v2 = initialState[2],
        v3 = initialState[3],
        v4 = initialState[4],
        v5 = initialState[5],
        v6 = initialState[6],
        v7 = initialState[7],
        v8 = initialState[8],
        v9 = initialState[9],
        v10 = initialState[10],
        v11 = initialState[11],
        v12 = initialState[12],
        v13 = initialState[13],
        v14 = initialState[14],
        v15 = initialState[15];

    assert(v0 == 0x61707865);
    assert(v1 == 0x3320646e);
    assert(v2 == 0x79622d32);
    assert(v3 == 0x6b206574);

    // -------------------------------------------------------------------------
    // Step 2: Do ROUNDS column/diagonal rounds
    //
    // We inlined the 'quarterRound' function because benchmarks showed
    // significant enough difference to non-inlined version.
    // -------------------------------------------------------------------------
    final roundsDividedByTwo = rounds ~/ 2;
    for (var i = 0; i < roundsDividedByTwo; i++) {
      // -------
      // Columns
      // -------
      v0 = uint32mask & (v0 + v4);
      v12 = rotateLeft32(v12 ^ v0, 16);
      v8 = uint32mask & (v8 + v12);
      v4 = rotateLeft32(v4 ^ v8, 12);
      v0 = uint32mask & (v0 + v4);
      v12 = rotateLeft32(v12 ^ v0, 8);
      v8 = uint32mask & (v8 + v12);
      v4 = rotateLeft32(v4 ^ v8, 7);

      v1 = uint32mask & (v1 + v5);
      v13 = rotateLeft32(v13 ^ v1, 16);
      v9 = uint32mask & (v9 + v13);
      v5 = rotateLeft32(v5 ^ v9, 12);
      v1 = uint32mask & (v1 + v5);
      v13 = rotateLeft32(v13 ^ v1, 8);
      v9 = uint32mask & (v9 + v13);
      v5 = rotateLeft32(v5 ^ v9, 7);

      v2 = uint32mask & (v2 + v6);
      v14 = rotateLeft32(v14 ^ v2, 16);
      v10 = uint32mask & (v10 + v14);
      v6 = rotateLeft32(v6 ^ v10, 12);
      v2 = uint32mask & (v2 + v6);
      v14 = rotateLeft32(v14 ^ v2, 8);
      v10 = uint32mask & (v10 + v14);
      v6 = rotateLeft32(v6 ^ v10, 7);

      v3 = uint32mask & (v3 + v7);
      v15 = rotateLeft32(v15 ^ v3, 16);
      v11 = uint32mask & (v11 + v15);
      v7 = rotateLeft32(v7 ^ v11, 12);
      v3 = uint32mask & (v3 + v7);
      v15 = rotateLeft32(v15 ^ v3, 8);
      v11 = uint32mask & (v11 + v15);
      v7 = rotateLeft32(v7 ^ v11, 7);

      // ---------
      // Diagonals
      // ---------
      v0 = uint32mask & (v0 + v5);
      v15 = rotateLeft32(v15 ^ v0, 16);
      v10 = uint32mask & (v10 + v15);
      v5 = rotateLeft32(v5 ^ v10, 12);
      v0 = uint32mask & (v0 + v5);
      v15 = rotateLeft32(v15 ^ v0, 8);
      v10 = uint32mask & (v10 + v15);
      v5 = rotateLeft32(v5 ^ v10, 7);

      v1 = uint32mask & (v1 + v6);
      v12 = rotateLeft32(v12 ^ v1, 16);
      v11 = uint32mask & (v11 + v12);
      v6 = rotateLeft32(v6 ^ v11, 12);
      v1 = uint32mask & (v1 + v6);
      v12 = rotateLeft32(v12 ^ v1, 8);
      v11 = uint32mask & (v11 + v12);
      v6 = rotateLeft32(v6 ^ v11, 7);

      v2 = uint32mask & (v2 + v7);
      v13 = rotateLeft32(v13 ^ v2, 16);
      v8 = uint32mask & (v8 + v13);
      v7 = rotateLeft32(v7 ^ v8, 12);
      v2 = uint32mask & (v2 + v7);
      v13 = rotateLeft32(v13 ^ v2, 8);
      v8 = uint32mask & (v8 + v13);
      v7 = rotateLeft32(v7 ^ v8, 7);

      v3 = uint32mask & (v3 + v4);
      v14 = rotateLeft32(v14 ^ v3, 16);
      v9 = uint32mask & (v9 + v14);
      v4 = rotateLeft32(v4 ^ v9, 12);
      v3 = uint32mask & (v3 + v4);
      v14 = rotateLeft32(v14 ^ v3, 8);
      v9 = uint32mask & (v9 + v14);
      v4 = rotateLeft32(v4 ^ v9, 7);
    }

    // -------------------------------------------------------------------------
    // Step 3: Addition (not done by Hchacha20)
    // -------------------------------------------------------------------------
    if (addAndXor) {
      state[si + 0] = (0xFFFFFFFF & (v0 + initialState[0])) ^ state[si + 0];
      state[si + 1] = (0xFFFFFFFF & (v1 + initialState[1])) ^ state[si + 1];
      state[si + 2] = (0xFFFFFFFF & (v2 + initialState[2])) ^ state[si + 2];
      state[si + 3] = (0xFFFFFFFF & (v3 + initialState[3])) ^ state[si + 3];
      state[si + 4] = (0xFFFFFFFF & (v4 + initialState[4])) ^ state[si + 4];
      state[si + 5] = (0xFFFFFFFF & (v5 + initialState[5])) ^ state[si + 5];
      state[si + 6] = (0xFFFFFFFF & (v6 + initialState[6])) ^ state[si + 6];
      state[si + 7] = (0xFFFFFFFF & (v7 + initialState[7])) ^ state[si + 7];
      state[si + 8] = (0xFFFFFFFF & (v8 + initialState[8])) ^ state[si + 8];
      state[si + 9] = (0xFFFFFFFF & (v9 + initialState[9])) ^ state[si + 9];
      state[si + 10] = (0xFFFFFFFF & (v10 + initialState[10])) ^ state[si + 10];
      state[si + 11] = (0xFFFFFFFF & (v11 + initialState[11])) ^ state[si + 11];
      state[si + 12] = (0xFFFFFFFF & (v12 + initialState[12])) ^ state[si + 12];
      state[si + 13] = (0xFFFFFFFF & (v13 + initialState[13])) ^ state[si + 13];
      state[si + 14] = (0xFFFFFFFF & (v14 + initialState[14])) ^ state[si + 14];
      state[si + 15] = (0xFFFFFFFF & (v15 + initialState[15])) ^ state[si + 15];
    } else {
      state[si + 0] = v0;
      state[si + 1] = v1;
      state[si + 2] = v2;
      state[si + 3] = v3;
      state[si + 4] = v4;
      state[si + 5] = v5;
      state[si + 6] = v6;
      state[si + 7] = v7;
      state[si + 8] = v8;
      state[si + 9] = v9;
      state[si + 10] = v10;
      state[si + 11] = v11;
      state[si + 12] = v12;
      state[si + 13] = v13;
      state[si + 14] = v14;
      state[si + 15] = v15;
    }
  }

  static void initializeChacha(
    Uint32List state, {
    required List<int> key,
    required List<int> nonce,
    int keyStreamIndex = 0,
  }) {
    if (key.length > 32) {
      throw ArgumentError('Invalid key');
    }
    if (nonce.length > 12) {
      throw ArgumentError('Invalid nonce');
    }
    state[0] = 0x61707865;
    state[1] = 0x3320646e;
    state[2] = 0x79622d32;
    state[3] = 0x6b206574;

    final stateByteData = ByteData.view(
      state.buffer,
      state.offsetInBytes,
      64,
    );

    //
    // Key
    //
    var stateBytesIndex = 16;
    for (var i = 0; i < key.length; i++) {
      stateByteData.setUint8(stateBytesIndex, key[i]);
      stateBytesIndex++;
    }

    //
    // Counter
    //
    state[12] = keyStreamIndex ~/ 64;

    //
    // Nonce
    //
    for (var i = 13; i < 16; i++) {
      state[i] = 0;
    }
    stateBytesIndex = 13 * 4;
    for (var i = 0; i < nonce.length; i++) {
      stateByteData.setUint8(stateBytesIndex, nonce[i]);
      stateBytesIndex++;
    }

    // In big endian platforms, convert little endian --> host endian
    if (Endian.host != Endian.little) {
      for (var i = 4; i < 12; i++) {
        state[i] = stateByteData.getUint32(4 * i, Endian.little);
      }
      for (var i = 13; i < 16; i++) {
        state[i] = stateByteData.getUint32(4 * i, Endian.little);
      }
    }
  }
}
